### PR TITLE
Rbac

### PR DESCRIPTION
**Pull request by Devjacky**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/auth/roles.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { SetMetadata } from '@nestjs/common';
+import { UserRoleEnum } from '../users/enum/user-role.enum';
+
+export const ROLES_KEY = 'roles';
+export const Roles = (...roles: UserRoleEnum[]) =>
```
Comment (kzkaneoka) [inline src/auth/roles.decorator.ts:5]: please change `(...roles: UserRoleEnum[])` into `(roles: UserRoleEnum[])`

File: src/auth/guards/roles.guard.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  canActivate(context: ExecutionContext): boolean {
+    const requiredRoles = this.reflector.getAllAndOverride<UserRoleEnum[]>(
+      ROLES_KEY,
+      [context.getHandler(), context.getClass()],
+    );
+    if (!requiredRoles) {
+      return true;
+    }
+    const { user } = context.switchToHttp().getRequest();
+    return requiredRoles.some((role) => user.roles?.includes(role));
```
Comment (kzkaneoka) [inline src/auth/guards/roles.guard.ts:19]: Let's change like this.
```ts
const { user } = context.switchToHttp().getRequest();
const { role } = await this.authService.findOne(user.id);
return requiredRoles.some((requiredRole) => requiredRole === role);
```

Also, please adjust async function and implement AuthService.findOne()

File: src/app.module.ts [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
 @Module({
   imports: [DatabaseModule, UsersModule, AuthModule],
   controllers: [AppController],
-  providers: [AppService],
+  providers: [
+    AppService,
+    {
+      provide: APP_GUARD,
+      useClass: JwtAuthGuard,
+    },
```
Comment (kzkaneoka) [inline src/app.module.ts:18]: i think we need to add `RolesGuard` as well.
```ts
{ provide: APP_GUARD, useClass: RolesGuard }
```

File: src/users/users.controller.ts [hunk 4]
```diff
@@ -13,6 +13,8 @@ import { CreateUserDto } from './dto/create-user.dto';
 import { UserModel } from './models/user.model';
 import { UsersService } from './users.service';
 import { UpdateUserDto } from './dto/update-user.dto';
+import { Roles } from 'src/auth/roles.decorator';
+import { UserRoleEnum } from './enum/user-role.enum';
 
 @Controller('users')
 export class UsersController {
```
Comment (kzkaneoka) [inline src/users/users.controller.ts:20]: Please decorate methods in this file with `@Roles` to control which endpoints are accessible for which user roles.
E.g)
```ts
@Get()
@Roles([UserRoleEnum.ADMIN, UserRoleEnum.USER])
findAll(): Promise<UserModel[]> {
```

**Review by kzkaneoka — COMMENTED** [review]
I've done reviews and please check them out. Thanks!

**Comment by Devjacky** [comment]
I have made the necessary changes, please review them. Thank you!

**Comment by kzkaneoka** [comment]
Hey, it looks like authorization needs to work on it. For example, I got an unauthorized error for `GET /users/:id`. This error happens no matter I logged in user role or admin role. Can you take care of it?

**Comment by kzkaneoka** [comment]
LGTM. It works from my end!

**Pull request merged**